File: Sorting/MergeSort.py
# Merge Sort

def mergeSort(nums):
    length = len(nums)
    if length <= 1:
        return nums
    # 将数组按照middle进行递归拆分
    mid = length // 2
    left = mergeSort(nums[:mid])
    right = mergeSort(nums[mid:])
    # 将其使用对两个有序数组进行排序的方法对其进行排序
    return merge(left, right)


def merge(left, right):
    res = []
    # 同时对两个数组的第一个位置进行比大小，将小的放入一个空数组，然后被放入空数组的那个位置的指针往后移一个
    # 然后继续和另外一个数组的上一个位置进行比较.
    while left and right:
        if left[0] >= right[0]:
            res.append(left[0])
            del left[0]
        else:
            res.append(right[0])
            del right[0]
    # 到最后任何一个数组先出栈完，就将另外i一个数组里的所有元素追加到新数组后面。
    if not left:
        res += right
    if not right:
        res += left

    return res


if __name__ == '__main__':
    a = [4, 7, 8, 3, 5, 9]
    print(mergeSort(a))

File: go_solution/sorting/mergeSort.go
package sorting

// 方法一：递归
func mergeSortRecursion(nums []int) []int {
	if len(nums) <= 1 {
		return nums
	}
	mid := len(nums) >> 1
	left := mergeSortRecursion(nums[:mid])
	right := mergeSortRecursion(nums[mid:])
	return merge(left, right)
}

func merge(listOne, listTwo []int) []int {
	var rlt []int
	idxOne, idxTwo := 0, 0
	for idxOne < len(listOne) && idxTwo < len(listTwo) {
		if listOne[idxOne] < listTwo[idxTwo] {
			rlt = append(rlt, listOne[idxOne])
			idxOne++
		} else {
			rlt = append(rlt, listTwo[idxTwo])
			idxTwo++
		}
	}
	// 如果listOne 和 listTwo 长度不一致，处理剩下的元素
	if idxOne < len(listOne) { // listOne 有遗留元素，直接append
		rlt = append(rlt, listOne[idxOne:]...)
	}
	if idxTwo < len(listTwo) { // listTwo 有遗留元素，直接append
		rlt = append(rlt, listTwo[idxTwo:]...)
	}
	return rlt
}
//--------------------------------
// 方法二：非递归
func mergeSortIteration(nums []int) {
	var level uint64
	for len(nums) > 1<<level {
		step := 1 << level
		l, r := 0, step
		for r < len(nums) {
			mergeInPlace(nums, l, r, step)
			l = r + step
			r = l + step
		}
		level++
	}
}

// store merger result in nums directly
func mergeInPlace(nums []int, left, right, step int) {
	lend, rend := step+left, step+right
	if len(nums) < rend {
		rend = len(nums)
	}
	curr, i, j := left, left, right
	for i < lend && j < rend {
		if nums[i] < nums[j] {
			i++
		} else {
			rotateForward(nums, j, curr)
			j++
			i++
		}
		curr++
	}
}

func rotateForward(nums []int, from, to int) {
	if from-to == 1 {
		nums[from], nums[to] = nums[to], nums[from]
	} else {
		val := nums[from]
		nums = append(nums[:to], append([]int{val}, append(nums[to:from], nums[from+1:]...)...)...)
	}
}
